lib/client.c

```diff
@@ -54,12 +56,173 @@ static SF_STATUS STDCALL
 _reset_connection_parameters(SF_CONNECT *sf, cJSON *parameters,
                              cJSON *session_info, sf_bool do_validate);
 
+static const char* query_status_names[] = {
+    "RUNNING",
+    "ABORTING",
+    "SUCCESS",
+    "FAILED_WITH_ERROR",
+    "ABORTED",
+    "QUEUED",
+    "FAILED_WITH_INCIDENT",
+    "DISCONNECTED",
+    "RESUMING_WAREHOUSE",
+    "QUEUED_REPAIRING_WAREHOUSE",
+    "RESTARTED",
+    "BLOCKED",
+    "NO_DATA"
+};
+
 /**
  * Validate partner application name.
  * @param application partner application name
  */
 sf_bool validate_application(const char *application);
 
+/**
+ * Helper function to get SF_QUERY_STATUS given the string representation
+ * @param query_status the string representation of the query status
+ */
+SF_QUERY_STATUS get_status_from_string(const char *query_status) {
+  if (query_status == NULL) {
+    return SF_QUERY_STATUS_NO_DATA;
+  }
+  int idx = 0, last = 0;
+  for (idx = 0, last = (int)SF_QUERY_STATUS_NO_DATA; idx <= last; ++idx) {
+    size_t len = strlen(query_status_names[idx]);
+    if (sf_strncasecmp(query_status_names[idx], query_status, len) == 0) {
+      return (SF_QUERY_STATUS)idx;
+    }
+  }
+  return SF_QUERY_STATUS_NO_DATA;
+}
+
+/**
+ * Get the metadata of the query
+ * @param sf the SF_CONNECT context
+ * @param query_id the query id
+ */
+char *get_query_metadata(SF_CONNECT *sf, const char *query_id) {
+  cJSON *resp = NULL;
+  cJSON *data = NULL;
+  cJSON *queries = NULL;
+  char *s_resp = NULL;
+  const char *error_msg;
+  size_t url_size = strlen(QUERY_MONITOR_URL) -2 + strlen(query_id) + 1;
+  char *status_query = (char*)SF_CALLOC(1, url_size);
+  sf_sprintf(status_query, url_size, QUERY_MONITOR_URL, query_id);
+
+  if (request(sf, &resp, status_query, NULL, 0, NULL, NULL,
+    GET_REQUEST_TYPE, &sf->error, SF_BOOLEAN_TRUE,
+    0, sf->retry_count, get_retry_timeout(sf),
+    NULL, NULL, NULL, SF_BOOLEAN_FALSE)) {
+
+    s_resp = snowflake_cJSON_Print(resp);
+    log_info("Here is JSON response:\n%s", s_resp);
+
+    data = snowflake_cJSON_GetObjectItem(resp, "data");
+
+    queries = snowflake_cJSON_GetObjectItem(data, "queries");
+    cJSON* query = snowflake_cJSON_GetArrayItem(queries, 0);
+
+    char *metadata = snowflake_cJSON_Print(query);
+    snowflake_cJSON_Delete(resp);
+    SF_FREE(s_resp);
+    SF_FREE(status_query);
+    return metadata;
+  }
+  SF_FREE(status_query);
+  log_trace("Error getting query metadata.");
+  return NULL;
+}
+
+/**
+ * Get the status of the query
+ * @param sf the SF_CONNECT context
+ * @param query_id the query id
+ */
+SF_QUERY_STATUS get_query_status(SF_CONNECT *sf, const char *query_id) {
```

Comment:
I think this would be something we expose to application? Please check JDBC async query usage to design what APIs should be expose to application. https://docs.snowflake.com/en/developer-guide/jdbc/jdbc-using#performing-an-asynchronous-query
I think better to use SF_STMT instance for argument.